File: src/apps/blueprint-gym/server/Domain/WorkoutTracker/Models/Workout.cs
using System.Collections.Generic;
using BlueprintGym.Domain.Constants.Enums;
using BlueprintGym.Domain.Core.Cosmos;
using BlueprintGym.Domain.Core.Interfaces;
using BlueprintGym.Domain.Core.Models;
using BlueprintGym.Domain.WorkoutTracker.Models.Containers;

namespace BlueprintGym.Domain.WorkoutTracker.Models
{
  public class Workout : CosmosEntity, ICosmosEntity
  {
    public override string PK => $"{this.GetType().Name}-{this.WorkoutId}";
    public string WorkoutId => this.Id;
    public string RegimenId { get; set; }
    public string WorkoutName { get; set; }
    public string WorkoutDescription { get; set; }
    public FitnessDifficulty Difficulty { get; set; }
    public IEnumerable<WorkoutExerciseAssignment> ExerciseAssignments { get; set; }
    public IEnumerable<ExerciseLink> BackupExercises { get; set; }
    public IEnumerable<string> WorkoutLabels { get; set; }
  }
}

File: apps/blueprint-gym/fitness-tracker/models/workout.go
package models

import "go.mongodb.org/mongo-driver/bson/primitive"

type Workout struct {
	ID                  primitive.ObjectID `json:"_id,omitempty" bson:"_id,omitempty"`
	RegimenId           string
	WorkoutName         string
	WorkoutDescription  string
	Difficulty          FitnessDifficulty
	ExerciseAssignments []WorkoutExerciseAssignment
	BackupExercises     []ExerciseLink
	WorkoutLabels       []string
}

type WorkoutRef struct {
	ID            primitive.ObjectID `json:"_id,omitempty" bson:"_id,omitempty"`
	RegimenId     string
	WorkoutName   string
	Difficulty    FitnessDifficulty
	MusclesWorked []string
}

type WorkoutEntry struct {
	ID            primitive.ObjectID `json:"_id,omitempty" bson:"_id,omitempty"`
	WorkoutId     string
	RegimenId     string
	WorkoutIndex  int
	TimeSubmitted primitive.DateTime
	SetGoals      []WorkoutSet
	SetEntries    []WorkoutSet
}

type WorkoutEntryRef struct {
	ID            primitive.ObjectID `json:"_id,omitempty" bson:"_id,omitempty"`
	WorkoutId     string
	RegimenId     string
	WorkoutIndex  int
	TimeSubmitted primitive.DateTime
	PointsEarned  int
}

type WorkoutExerciseAssignment struct {
	Order                int
	SetIdentifier        string
	SpecialSetIdentifier string
	ExerciseId           string
	ExerciseName         string
	HeavyAim             ExerciseAimInfo
	ConditionedAim       ExerciseAimInfo
	DurableAim           ExerciseAimInfo
	SpecializedSetType   SpecializedSetType
	MusclesWorked        []string
}

type ExerciseAimInfo struct {
	AimBonusCutoff int
	ExerciseAim    ExerciseAim
}

type WorkoutSet struct {
	EntryId       string
	SetIdentifier string
	Weight        int
	Reps          int
}
